### ISSUE TITLE
Question: Ranges Variable and Walk Forward Optimization

### ISSUE THREAD

**Issue by CMobley7** [jmrichardson/tuneta#20]
To Whom It May Concern,

I see in https://github.com/jmrichardson/tuneta/blob/main/examples/tune_all.py, that you are tuning all indicator over a period of 10 years of daily data given the next day's returns. What does `ranges=[(2, 30)]` do in this instance? I see that ranges is defined here, https://github.com/jmrichardson/tuneta/blob/main/tuneta/tune_ta.py#L44. It specifies that it is a `parameter search space`. It appears to be used to define a low and high here, https://github.com/jmrichardson/tuneta/blob/main/tuneta/tune_ta.py#L82, which then appears to be used as a parameter in an `Optuna` trial here, https://github.com/jmrichardson/tuneta/blob/main/tuneta/tune_ta.py#L125. Unfortunately, it's still not clear to me what this parameter does. My apologies for my ignorance.

I'm using `pandas-ta` on minute data. So, I need to adjust the default setting for most of indicators I'm using. I had hope to use this library to get these settings using a `walk forward optimization strategy`, which ensures that I'm not overfitting my data and/or getting stuck in a local minima. Just for completeness as I wasn't aware of this a few months back, but a walk forward optimization is defined here, https://en.wikipedia.org/wiki/Walk_forward_optimization and https://www.youtube.com/watch?v=GowmmrSMw9I, and shown in code https://github.com/polakowo/vectorbt/blob/master/examples/WalkForwardOptimization.ipynb on `VectorBT`, as well as visualized here, http://www.adaptivetradingsystems.com/blog/modeling_sofware/walk-forward-simulations-in-synergy/. 

Is it possible to use `tuneta` to adjust the setting for `pandas-ta` indicators in this way. If not, could we discuss how to make this possible?

Thanks for the hard work on this library, as well as your time and attention to this matter. I hope that this message finds you well and that you have a great week. God bless.

Very Respectfully,
CMobley7

**Comment by jmrichardson** [comment]
Hi @CMobley7 ,

The ranges parameter simply defines the lower and upper limit of the indicator parameter(s) values to test.  For example, A ranges=[(2, 30]) for pta.RSI would result in the RSI indicator being tuned with a Length parameter between 2 and 30.  Because ranges is a list of tuples, you could specify something like ranges=[(2,30), (31,90), (91,180)] which would result in the RSI indicator being optimized 3 times for each range  (ie between 2 and 30, between 31 and 90, and 91 and 180).  The final data set would contain all 3 RSI indicators tuned between each range.  Some indicators perform well on multiple ranges which is why this was implemented.  Other indicators may only perform well on a particular popular range.  

WFO is not included as a feature in TuneTA.  However, you could simply create a loop on each data set split and use tuneta for each.

Hope this helps.

**Comment by CMobley7** [comment]
Wow @jmrichardson ,

Thank for the quick response! I appreciate it. Understood, regarding ranges and sorry for the plethora of questions. So, would ranges be used to set the following parameters, https://github.com/jmrichardson/tuneta/blob/main/tuneta/config.py#L5-L16, if they exist for a particular indicator? If so, and an indicator has more than 1 of those parameters, would each be tuned simultaneously and independently? Also, if I'm doing multiple indicators, such as in https://github.com/jmrichardson/tuneta/blob/main/examples/tune_all.py, then would they all be tuned simultaneously and independently, would each indicators be tuned sequentially, or would each of the parameters from https://github.com/jmrichardson/tuneta/blob/main/tuneta/config.py#L5-L16 be tuned simultaneously but not independently and the optimization would just find a single value of `lower_length` that would lead to a maximization of distance correlation across all the indicators that accept that parameter for the training/validation set? Again, sorry for so many question, but thank you for being responsive and willing to answer them.

**Comment by jmrichardson** [comment]
Hi @CMobley7 ,

No problem with the questions at all.  Yes, exactly, the parameters you highlighted are used to tune if they exist.  

When an indicator has multiple parameters, all are tuned as a set using Optuna.  In other words, Optuna will try a set of parameters (all within the range), then another set, iteratively using its algorithm to choose a set that optimizes it according to the target (ie next day return).  Each indicator is tuned independently and done in parallel/simultaneously using joblib.  To be clear, each indicator is tuned independently from any other indicator and NOT in your example lower_length across all.  I haven't even considered that as an option and wondering if that is of value?

**Comment by CMobley7** [comment]
Thanks again! The way you're doing it currently is probably the best way to go about it. I wouldn't have wanted `lower_length` to have the same value across all indicators. Each indicator is different; so, it makes sense to have their values tuned independently. Thanks for the great library. I look forward to using it more in depth over the coming weeks. 

Would you be open to moving to from `pandas-ta's` `main` branch to it's `development` branch. `twopirllc's` `pandas-ta` is amazing and the `development` branch is now 166 commits ahead of main and has several new indicators, lots of bug fixes, speed ups, etc. I've been using it for months at this point and it's pretty stable even with active development. `twopirllc` is really good about testing changes before merging them. Not sure when he'll update master. Also, in regards to `ichimoku`, https://github.com/jmrichardson/tuneta/blob/main/tuneta/config.py#L69, and `dpo`, https://github.com/jmrichardson/tuneta/blob/main/tuneta/config.py#L155, both of these have a variable that prevent data leakage, https://github.com/twopirllc/pandas-ta/blob/development/pandas_ta/overlap/ichimoku.py#L55-L56 and https://github.com/twopirllc/pandas-ta/blob/development/pandas_ta/overlap/ichimoku.py#L55-L56, respectively. Would you be open to adding these back in with those parameters set?

**Comment by jmrichardson** [comment]
It would be 500 trials per indicator.  

Pandas-ta is a great library!  I do not put any versioning controls on any of the packages (see requirements.txt).  So you are free to use the dev version to get the speed improvements.  However, I don't want to force everyone to the dev branch on install which can cause versioning problems when other packages may depend on the main branch.  There may be some additional indicators that are not in the main branch yet and if I add them I think it will cause some errors when indicators are not specifically defined.  Unfortunately, the dpo and ichimoku require a specific parameter to avoid data leakage which tuneta cannot provide.  That is, the parameters are all given to Optuna and no indicators have a specifically defined set of parameters.  I would have to think about how to implement this...  If you want to implement with a PR, please feel free :)

**Comment by CMobley7** [comment]
@jmrichardson , thanks again for your quick and thorough responses! I might have some time this upcoming weekend or next to submit a PR for this.

**Comment by CMobley7** [comment]
The change I'd need to make is approximately 4-5 lines across 2 files. I should submit by Sunday evening. I'm still familiarizing myself with the entire code base. Would you care if I ran isort, import sorter, and black, code formatter, on these files too?

**Comment by CMobley7** [comment]
Also, @jmrichardson  what are your thoughts on @wouldayajustlookatit's changes in https://github.com/jmrichardson/tuneta/compare/main...wouldayajustlookatit:wouldya-dev. I'm personally a fan.

**Comment by jmrichardson** [comment]
> The change I'd need to make is approximately 4-5 lines across 2 files. I should submit by Sunday evening. I'm still familiarizing myself with the entire code base. Would you care if I ran isort, import sorter, and black, code formatter, on these files too?

@CMobley7 Perfect!  Sounds good and feel free to improve the code in any way.  Unfortunately, I don't have unit testing enabled on this repo but hopefully you can verify at minimum the examples to make sure they all work as expected.

**Comment by jmrichardson** [comment]
@CMobley7 

It looks like @wouldayajustlookatit repo is based on an earlier version prior to a major code update to support different enhancements such as distance correlation, multiple equities, cluster based selection of parameters, etc.  Unfortunately, I can't merge the updates because the code base is significantly different between both.  However, if you see any updates that make sense, feel free to add/test them with your PR.  Happy to merge any improvements!
